File: hw-kvsrv/src/lock.c
#include "src/lock.h"

#include <stdlib.h>
#include <string.h>

lock_t *lock_new(kv_client_t *client, const char *lockname) {
    lock_t *l = calloc(1, sizeof(lock_t));
    l->client = client;
    // TODO: initialize lock identity and version.
    return l;
}

void lock_free(lock_t *l) {
    free(l);
}

void lock_acquire(lock_t *l) {
    // TODO: acquire the lock through the KV client.
}

void lock_release(lock_t *l) {
    // TODO: release the lock if this client owns it.
}

File: hw-kvsrv-rs/src/lock.rs
// Part 2: Distributed Lock with KV Server
// Part 4: Distributed Lock with Raft

use std::sync::Arc;

use crate::client::*;

pub struct Lock {
    client: Arc<dyn KvClient>,
    lockname: String,
    // TODO: add any other lock client state.
}

impl Lock {
    pub fn new(client: Arc<dyn KvClient>, lockname: &str) -> Self {
        // TODO: initialize lock state.
        todo!()
    }

    pub async fn acquire(&mut self) {
        // TODO: acquire the lock through the KV client.
        todo!()
    }

    pub async fn release(&mut self) {
        // TODO: release the lock if this client owns it.
        todo!()
    }
}
